chain/epoch_manager/src/reward_calculator.rs

```diff
@@ -83,47 +88,127 @@ impl RewardCalculator {
+                        .expected
+                        > 0
+                    {
+                        *stake
+                    } else {
+                        0
+                    }
+                })
+                .sum();
+            let cp_total_stake: Balance = validator_stake
+                .iter()
+                .map(|(account_id, stake)| {
+                    if validator_block_chunk_stats
+                        .get(account_id)
+                        .unwrap_or_else(|| panic!("{} is not a validator", account_id))
+                        .chunk_stats
+                        .expected
+                        > 0
+                    {
+                        *stake
+                    } else {
+                        0
+                    }
+                })
+                .sum();
+            let block_reward_percentage = to_big_rational(self.block_reward_percentage);
+            let chunk_reward_percentage =
+                to_big_rational(Rational::from_integer(1) - self.block_reward_percentage);
+            for (account_id, stats) in validator_block_chunk_stats {
+                let block_uptime_mult = calc_uptime_mult(&stats.block_stats);
+                let chunk_uptime_mult = calc_uptime_mult(&stats.chunk_stats);
                 let stake = *validator_stake
                     .get(&account_id)
                     .unwrap_or_else(|| panic!("{} is not a validator", account_id));
-                // Online reward multiplier is min(1., (uptime - online_threshold_min) / (online_threshold_max - online_threshold_min).
-                let online_max_numer = U256::from(*self.online_max_threshold.numer() as u64);
-                let online_max_denom = U256::from(*self.online_max_threshold.denom() as u64);
-                let online_numer =
-                    online_max_numer * online_min_denom - online_min_numer * online_max_denom;
-                let mut uptime_numer = (average_produced_numer * online_min_denom
-                    - online_min_numer * average_produced_denom)
-                    * online_max_denom;
-                let uptime_denum = online_numer * average_produced_denom;
-                // Apply min between 1. and computed uptime.
-                uptime_numer =
-                    if uptime_numer > uptime_denum { uptime_denum } else { uptime_numer };
-                (U256::from(epoch_validator_reward) * uptime_numer * U256::from(stake)
-                    / uptime_denum
-                    / U256::from(total_stake))
-                .as_u128()
-            };
-            res.insert(account_id, reward);
-            epoch_actual_reward += reward;
+                let block_stake_ratio = BigRational::new(stake.into(), bp_total_stake.into());
+                let chunk_stake_ratio = BigRational::new(stake.into(), cp_total_stake.into());
+                let reward = (block_stake_ratio * block_uptime_mult * &block_reward_percentage
```

Comment:
I'm a little worried about the 'rounding errors' here ..
would it be possible that epoch_actual_reward ends up larger than epoch_total_reward?